### DISCUSSION TITLE
FormatterAttribute build error and wrong auto generation

### DISCUSSION THREAD

**Discussion by cengkaplan33** [serenity-is/Serenity#7283] [Help]
After I opened new project. While it is empty , I tried to write a new formatter.  So I checked sample codes  then I write  a formatter. But project is not build after all.

I know it is a simple task, but I cannot worked it.

I read the another topics. It may caused from node.js . Then I removed node.js and installed it but nothing changed.
by the way sergen is working. 
When I closed my code. Everything is fine.

and Another suspicion is wheren I changed    namespace  EmployeeFormatterAttribute.cs 
from Kaptan.Administration to Kaptan.Administration.Formatter . Evertthing is fine but I added step6 auto generated code and Image .
There is no my GenderProperty in generated code.
Something goes wrong.

step1
.cs file code is;
namespace Kaptan.Administration;

public partial class EmployeeFormatterAttribute : CustomFormatterAttribute
{
    public const string Key = "Kaptan.Administration.EmployeeFormatter";

    public EmployeeFormatterAttribute()
        : base(Key)
    {
    }

    public string GenderProperty
    {
        get { return GetOption<string>("genderProperty"); }
        set { SetOption("genderProperty", value); }
    }
}

step2
.ts file code is ;
import { Decorators, Formatter, IInitializeColumn, isTrimmedEmpty } from "@serenity-is/corelib";
import { Column, FormatterContext  } from "@serenity-is/sleekgrid";

@Decorators.registerFormatter('Kaptan.Administration.EmployeeFormatter', [IInitializeColumn])
export class EmployeeFormatter implements Formatter {
    constructor(public readonly props: { genderProperty?: string } = {}) {
        this.props ??= {};
    }

    format(ctx: FormatterContext): string{
        if (!this.props.genderProperty || isTrimmedEmpty(ctx.value))
            return ctx.value;

        let propertyValue = ctx.item[this.props.genderProperty];
        return '<i class="fa fa-female">' + propertyValue + ' - ' + ctx.value + '</i> ';
    }

    public initializeColumn(column: Column) {
        column.referencedFields = column.referencedFields || [];
        if (this.props.genderProperty)
            column.referencedFields.push(this.props.genderProperty);
    }
}
step3.
attribute usage code is ;

step4
error is;

Error	CS0111	Type 'EmployeeFormatterAttribute' already defines a member called 'EmployeeFormatterAttribute' with the same parameter types	
...\Kaptan.Web\Serenity.Pro.Coder\Serenity.Pro.Coder.ClientTypesSourceGenerator\Administration.EmployeeFormatterAttribute.generated.cs	13	Kaptan.Web	Active
Error	CS0102	The type 'EmployeeFormatterAttribute' already contains a definition for 'Key'	...\Kaptan.Web\Serenity.Pro.Coder\Serenity.Pro.Coder.ClientTypesSourceGenerator\Administration.EmployeeFormatterAttribute.generated.cs	11	Kaptan.Web	Active
Error	CS0229	Ambiguity between 'EmployeeFormatterAttribute.Key' and 'EmployeeFormatterAttribute.Key'	...\Kaptan.Web\Modules\Administration\User\EmployeeFormatterAttribute.cs

Build message is;
![image](https://github.com/user-attachments/assets/33621fc0-b544-4bfd-a874-e8594856e70e)

Step5;
usage code is ;
namespace Kaptan.Administration.Columns;

[ColumnsScript("Administration.User")]
[BasedOnRow(typeof(UserRow), CheckNames = true)]
public class UserColumns
{
  ..

    [Width(32)]
    [EmployeeFormatterAttribute]
    public string ImpersonationToken { get; set; }
...
}

usage image is ;

![image](https://github.com/user-attachments/assets/09bf933e-07a3-4d5c-9017-fc6f9c0861d0)

step6
auto generation code is ;

This file is auto-generated by the generator 'Serenity.Pro.Coder.ClientTypesSourceGenerator' and cannot be edited. 
...
...
.
namespace Kaptan.Administration
{
    public partial class EmployeeFormatterAttribute : CustomFormatterAttribute
    {
        public const string Key = "Kaptan.Administration.EmployeeFormatter";

        public EmployeeFormatterAttribute()
            : base(Key)
        {
        }
    }
}

as a image;
![image](https://github.com/user-attachments/assets/5146fc4a-e828-487b-84b3-028e69d75b8b)

**Comment by VictorTomaili** [comment]
You don't need to write formatter attribute. Just write the formatter in typescript and build project, attribute should be generated automatically. Then use the auto generated attribute.

**Reply by cengkaplan33** [reply]
Ok , I deleted  EmployeeFormatterAttribute .cs,
Buid is ok but how can I set GenderProperty? 

I will use GenderProperty for gender info is in which column? 
For example user info is coming from join .  So I will set GenderProperty  as  "UserGender" so , in grid I hope to take value from "item"


auto generated code is ; 
namespace Kaptan.Administration
{
    public partial class EmployeeFormatterAttribute : CustomFormatterAttribute
    {
        public const string Key = "Kaptan.Administration.EmployeeFormatter";

        public EmployeeFormatterAttribute()
            : base(Key)
        {
        }
    }
}

**Reply by cengkaplan33** [reply]
@VictorTomaili  I hope u noticed, my sample code is your employeeFormmatter which is in northwind demo.
but it is not working for me.

ts file
![image](https://github.com/user-attachments/assets/83bcd453-8c6d-4c42-b4f6-8156890ec10f)

generated code; and there is GenderProperty. Why is it not working for me ?
![image](https://github.com/user-attachments/assets/e82dde62-4ee3-4326-9f70-fff43760928c)

**Comment by VictorTomaili** [comment]
As you can see below, I added `EmployeesFormatter` (added `s` to change the name of the formatter, and renamed `genderProperties` to make difference) 

I wrote code in UserDialog.ts file but it better on separated file. 
After build project I can see the `EmployeesFormatterAttribute` in path `\Imports\ClientTypes\Administration.EmployeesFormatterAttribute.cs`. The property is generated from typescript constructor parameters `public readonly props: { genderProperties?: string } = {}`
Then I can write `EmployeesFormatter(GenderProperties = "MyProperty")`
Don't forget the `GenderProperties` is a property, not a constructor parameter. So you need to explicitly write the name of the property.
If it is not works in your project, check the serenity version and any other errors while building the project. 

![image](https://github.com/user-attachments/assets/595cf113-26c0-4ce6-a3a2-d0c57975eda2)

**Reply by cengkaplan33** [reply]
Code is exactly same but my generation is wrong.


namespace Kaptan.Administration
{
    public partial class EmployeeFormatterAttribute : CustomFormatterAttribute
    {
        public const string Key = "Kaptan.Administration.EmployeeFormatter";

        public EmployeeFormatterAttribute()
            : base(Key)
        {
        }
    }
}

my version is 8.1.5 : how can I downgrade by serenity version at now or can I ?
my .csproj is
 
![image](https://github.com/user-attachments/assets/b720fbfa-056c-4f25-bc5f-827b7904f1ad)



 Project Sdk="Microsoft.NET.Sdk.Web" 
   PropertyGroup 
     TargetFramework net8.0 /TargetFramework 
     RootNamespace Kaptan /RootNamespace 
     TypeScriptCompileBlocked true /TypeScriptCompileBlocked 
     TypeScriptNoEmitOnError false /TypeScriptNoEmitOnError 
     DotNetSergen dotnet sergen /DotNetSergen 
   /PropertyGroup 
   ItemGroup 
     TypeScriptCompile Remove="node_modules\**" / 
     EmbeddedResource Include="Migrations\**\*.sql" / 
     Content Update=".eslintrc.json;libman.json;package.json;package-lock.json;sergen.json;tsconfig.json;tslint.json;Modules\tsconfig.json" CopyToOutputDirectory="Never" CopyToPublishDirectory="Never" / 
     Using Include="Microsoft.AspNetCore.Mvc;Microsoft.Extensions.Options;System;System.Collections.Generic;System.ComponentModel;System.Linq;System.Text;System.Threading" / 
     Using Include="Serenity;Serenity.Abstractions;Serenity.ComponentModel;Serenity.Data;Serenity.Data.Mapping;Serenity.Extensions;Serenity.Pro.Extensions;Serenity.Services;Serenity.Web" / 
     Using Include="System.Data.IDbConnection" Alias="IDbConnection" / 
   /ItemGroup 
   ItemGroup 
     PackageReference Include="EPPlus" Version="4.5.3.3" / 
     PackageReference Include="FluentMigrator.Runner" Version="3.3.2" / 
     PackageReference Include="Microsoft.Data.Sqlite" Version="8.0.0" / 
     PackageReference Include="Microsoft.Data.SqlClient" Version="5.1.2" / 
     PackageReference Include="Microsoft.TypeScript.MSBuild" Version="5.2.2" PrivateAssets="all" IncludeAssets="runtime; build; native; contentfiles; analyzers" / 
   /ItemGroup 
   ItemGroup 
     PackageReference Include="Serenity.Assets" Version="8.1.5" / 
     PackageReference Include="Serenity.Corelib" Version="8.1.5" / 
     PackageReference Include="Serenity.Net.Web" Version="8.1.5" / 
     PackageReference Include="Serenity.Extensions" Version="8.1.5" / 
     PackageReference Include="Serenity.Pro.Coder" Version="8.1.5" / 
     PackageReference Include="Serenity.Pro.Extensions" Version="8.1.5" / 
     PackageReference Include="Serenity.Pro.Theme" Version="8.1.5" / 
     PackageReference Include="Serenity.Pro.DataAuditLog" Version="8.1.5" / 
     PackageReference Include="Serenity.Pro.EmailQueue" Version="8.1.5" / 
   /ItemGroup 
   Target Name="NpmInstall" BeforeTargets="BeforeBuild" Condition="!Exists('node_modules\.bin\esbuild')" 
     Message Text="Installing NPM packages..." Importance="high" / 
     Exec Command="npm install" / 
   /Target 
   Target Name="RunTSBuild" BeforeTargets="BeforeBuild;GenerateStaticWebAssetsManifest" DependsOnTargets="RestoreNodeTypes;NpmInstall" Condition="'$(SkipTSBuild)' != 'true'" 
     Message Text="Executing TypeScript compiler for $(ProjectName)..." Importance="high" / 
     ItemGroup 
       CleanedEsmFiles Include="wwwroot\esm\**" / 
     /ItemGroup 
     Exec Command="node ./tsbuild.js" ContinueOnError="true" / 
     ItemGroup 
       CleanedEsmFiles Remove="wwwroot\esm\**" / 
       Content Remove="@(CleanedEsmFiles)" / 
       Content Include="wwwroot\esm\**\*.js;wwwroot\esm\**\*.css;wwwroot\esm\**\*.map" Exclude="@(Content)" CopyToOutputDirectory="Never" CopyToPublishDirectory="PreserveNewest" ExcludeFromSingleFile="true" / 
     /ItemGroup 
   /Target 
   Target Name="TransformMvcClientTypes" BeforeTargets="BeforeBuild" 
     Exec Command="dotnet tool restore" ContinueOnError="true" / 
     Exec Command="$(DotNetSergen) mvct" ContinueOnError="true" / 
   /Target 
   Target Name="TransformServerTypings" AfterTargets="AfterBuild" 
     Exec Command="$(DotNetSergen) servertypings" ContinueOnError="true" / 
   /Target 
   Target Name="TranformRunTSBuild" AfterTargets="AfterBuild" DependsOnTargets="RestoreNodeTypes;NpmInstall;TransformServerTypings" Condition="'$(SkipTSBuild)' != 'true'" 
     Message Text="Executing TypeScript compiler for $(ProjectName)..." Importance="high" / 
     Exec Command="node ./tsbuild.js" ContinueOnError="true" / 
   /Target 
   ProjectExtensions  VisualStudio  UserProperties NpmRestoreOnProjectOpen="True" /  /VisualStudio  /ProjectExtensions 
 /Project

**Reply by VictorTomaili** [reply]
I used latest version, ensure the project is built without any error, then check the attribute and add attribute to columns.cs file.

**Reply by cengkaplan33** [reply]
I hope to find something wrong for fix bu everything is ok.
build is ok too.

![image](https://github.com/user-attachments/assets/a3b228cc-644b-4555-a0ca-af6360479e64)

Meantime , I installed  Serene  . then I created new project with serene instead of StartSharp. 
I copied formatter to new project.  After build everything is ok. FormatterAttribute created GenderPropery is in there.  
I dont know but I guess there is a problem in StartSharp .
I looked up everywhere for solution but I am helpless now.

serene output;
![image](https://github.com/user-attachments/assets/552facaa-0b53-4729-b84f-e849b8c51260)

**Reply by cengkaplan33** [reply]
Can it be a clue? 
In StartSharp project there is no Import/ClientTypes folder. But In Serene Application folder is  exists.

![image](https://github.com/user-attachments/assets/e6a15d4b-9c68-4712-b62a-4f93905f9013)

**Reply by VictorTomaili** [reply]
If you are a StartSharp member, you need to ask StartSharp related questions in the StartSharp repository issues or in emails. There are many difference between serene and StartSharp. And no, the image is not meaningful about the issue. This is because of the Pro.Coder

**Reply by cengkaplan33** [reply]
Thanks @VictorTomaili  . I understood you but I actually stated the problem here so that it would contribute to the development of the Serenity.

If there is no problem with my computer or code , then there is a bug exists.

**Reply by VictorTomaili** [reply]
There is no bug, I tried and there are no problem about that. It probably something related to your environment or project.

**Reply by cengkaplan33** [reply]
Thank you @VictorTomaili  . I changed my code to old sample then it is worked.
If it helps anyone I am sharing my code.

import { Decorators, Formatter, htmlEncode, isTrimmedEmpty } from '@serenity-is/corelib';
import { Column, FormatterContext } from "@serenity-is/sleekgrid"

@Decorators.registerFormatter('BayrakFormatter')
export class BayrakFormatter implements Formatter {
    constructor() { }

    format(ctx: FormatterContext) {

        var text = htmlEncode(ctx.value);

        if (!this.bayrakProperty || isTrimmedEmpty(ctx.value))
            return text;

        console.log("deneme");

        var bayrak = ctx.item[this.bayrakProperty] as string;
        console.log(bayrak);

        //return "<a class='f16'>  <span class= ' flag " + bayrak.toLowerCase() + "'></span>  " + text + " </a>";
        return "<a class='f16'>  <span class= ' flag tr'></span>  " + text + " </a>";


    }

    @Decorators.option()
    public bayrakProperty: string;

    public initializeColumn(column: Column) {
        column.referencedFields = column.referencedFields || [];
        if (this.bayrakProperty)
            column.referencedFields.push(this.bayrakProperty);
    }
}

**Reply by VictorTomaili** [reply]
This is because your serenity version is old.
